### PR TITLE
[ENH] Add MRIQC client for feature crowdsourcing

### PR DESCRIPTION
**Pull request by Shotgunosine**
I've added MRIQC client code for uploading qc metrics to https://github.com/nih-fmrif/mriqc_db_endpoint which is currently running on an AWS server at 34.201.213.252.

### CODE REVIEW

**Comment by chrisfilo** [comment]
Great stuff! Thank you! Closes #383.

**Comment by oesteban** [comment]
This is very cool, thanks!

**Comment by chrisfilo** [comment]
I had some discussions with @rwblair and @394822740 and we thought adding the following features would be good:
- [ ] Flattening the payload before submission. This would eliminate the "metadata" subdictionary and make filtering the database via API easier.
- [ ] Whitelisting the fields from metadata. We want to avoid 1) accidental disclosure of personal information 2) accidental uploads of large amount of data. The allowed fields should be limited to:
- ContrastBolusIngredient
- RepetitionTime
- TaskName
- Manufacturer
- ManufacturersModelName
- MagneticFieldStrength
- DeviceSerialNumber
- SoftwareVersions
- HardcopyDeviceSoftwareVersion
- ReceiveCoilName
- GradientSetType
- MRTransmitCoilSequence
- MatrixCoilMode
- CoilCombinationMethod
- PulseSequenceType
- PulseSequenceDetails
- NumberShots
- ParallelReductionFactorInPlane
- ParallelAcquisitionTechnique
- PartialFourier
- PartialFourierDirection
- PhaseEncodingDirection
- EffectiveEchoSpacing
- TotalReadoutTime
- Timing Parameters
- EchoTime
- InversionTime
- SliceTiming
- SliceEncodingDirection
- NumberOfVolumesDiscardedByScanner
- NumberOfVolumesDiscardedByUser
- DelayTime
- FlipAngle
- MultibandAccelerationFactor
- Instructions
- TaskDescription
- CogAtlasID
- CogPOID
- InstitutionName
- InstitutionAddress
- ConversionSoftware
- ConversionSoftwareVersion
- md5sum
- modality
- mriqc_pred
- software
- subject_id
- version

@Shotgunosine if you could add this to this PR that would be absolutely fantastic and mega cool, but if not let us know and someone else will take over.

**Comment by Shotgunosine** [comment]
@chrisfilo As of tomorrow I'll be out of town with limited communication till May 8th, but I think I've addressed the changes you wanted.

File: mriqc/reports/utils.py [hunk 1]
```diff
@@ -109,6 +109,81 @@ def read_report_snippet(in_file):
...
+            in_data = load(h)
+        #metadata whitelist
+        whitelist = [u"ContrastBolusIngredient",u"RepetitionTime",u"TaskName",u"Manufacturer",
+                     u"ManufacturersModelName",u"MagneticFieldStrength",u"DeviceSerialNumber",
+                     u"SoftwareVersions",u"HardcopyDeviceSoftwareVersion",u"ReceiveCoilName",
+                     u"GradientSetType",u"MRTransmitCoilSequence",u"MatrixCoilMode",
+                     u"CoilCombinationMethod",u"PulseSequenceType",u"PulseSequenceDetails",
+                     u"NumberShots",u"ParallelReductionFactorInPlane",u"ParallelAcquisitionTechnique",
+                     u"PartialFourier",u"PartialFourierDirection",u"PhaseEncodingDirection",
+                     u"EffectiveEchoSpacing",u"TotalReadoutTime",u"Timing Parameters",
```
Comment (chrisfilo) [inline mriqc/reports/utils.py]: I included "Timing Parameters", by accident. It should not be on this list. Sorry!

**Comment by Shotgunosine** [comment]
I can try to fix this, but if one of you has a chance to take a look at the error and point me in the right direction, I'd appreciate it.

**Comment by chrisfilo** [comment]
Nothing jumps out to me - maybe @oesteban can help.

**Comment by Shotgunosine** [comment]
As far as I can tell this is failing because the CircleCI chain for this pull request has cached an out of date data directory. @oesteban or @chrisfilo, can you go to https://circleci.com/gh/poldracklab/mriqc/1720#tests and rebuild it without cache?

**Comment by Shotgunosine** [comment]
At the moment this branch fails because the hashes for the bold images don't match the values in the test file. As far as I can tell though, the difference in hashes is not being caused by this branch. 

We've built a clone of the master branch with the workflow versions from the prepare release branch (https://circleci.com/gh/nih-fmrif/mriqc/11#tests/containers/1) and checked those hashes against the hashes produced when we build the client branch with the same requirement versions (https://circleci.com/gh/nih-fmrif/mriqc/9#tests/containers/1). We get the same hashes for our master and client builds, so I think the client code would be passing the tests if the testing hashes were correct.

File: mriqc/bin/mriqc_run.py [hunk 2]
```diff
@@ -300,6 +313,10 @@ def main():
         if n_group_reports == 0:
             raise Exception("No data found. No group level reports were generated.")
 
+    if settings['no_sub'] == False:
```
Comment (oesteban) [inline mriqc/bin/mriqc_run.py]: Just `if not settings['no_sub']:`

File: mriqc/reports/utils.py [hunk 3]
```diff
@@ -109,6 +109,78 @@ def read_report_snippet(in_file):
...
+    report_log.setLevel(logging.INFO)
+
+    if no_sub == True:
+        report_log.info('QC metrics were not uploaded because --no_sub or --testing options were set.')
+        r = "No Response"
+    else:
+        with open(in_iqms, 'r') as h:
+            in_data = load(h)
+        #metadata whitelist
+        whitelist = [u"ContrastBolusIngredient",u"RepetitionTime",u"TaskName",u"Manufacturer",
```
Comment (oesteban) [inline mriqc/reports/utils.py]: MRIQC is now Python 3 only. Please remove the incompatible `u-` prefixes to these strings.

File: mriqc/reports/utils.py [hunk 4]
```diff
@@ -109,6 +109,78 @@ def read_report_snippet(in_file):
...
+                     u"PartialFourier",u"PartialFourierDirection",u"PhaseEncodingDirection",
+                     u"EffectiveEchoSpacing",u"TotalReadoutTime",
+                     u"EchoTime",u"InversionTime",u"SliceTiming",u"SliceEncodingDirection",
+                     u"NumberOfVolumesDiscardedByScanner",u"NumberOfVolumesDiscardedByUser",
+                     u"DelayTime",u"FlipAngle",u"MultibandAccelerationFactor",u"Instructions",
+                     u"TaskDescription",u"CogAtlasID",u"CogPOID",u"InstitutionName",
+                     u"InstitutionAddress",u"ConversionSoftware",u"ConversionSoftwareVersion",
+                     u"md5sum",u"modality",u"mriqc_pred",u"software",u"subject_id",u"version"]
+        # flatten data
+        data = {k:v for k,v in in_data.items() if k != 'metadata'}
```
Comment (oesteban) [inline mriqc/reports/utils.py]: Missing `list` (Py3 compat):
```
data = {k:v for k,v in list(in_data.items()) if k != 'metadata'}
```

File: mriqc/reports/utils.py [hunk 5]
```diff
@@ -109,6 +109,78 @@ def read_report_snippet(in_file):
...
+        # Filter Metadata values that aren't in whitelist
+        try:
+            data.update({k:v for k,v in in_data['metadata'].items() if k in whitelist})
```
Comment (oesteban) [inline mriqc/reports/utils.py]: Please make `list` of `items()`

File: mriqc/reports/utils.py [hunk 6]
```diff
@@ -109,6 +109,78 @@ def read_report_snippet(in_file):
...
+        except KeyError:
+            pass
+        # Preemptively adding code to handle settings
+        try:
+            data.update({k:v for k,v in in_data['settings'].items() if k in whitelist})
```
Comment (oesteban) [inline mriqc/reports/utils.py]: Please make `list` of `items()`

File: mriqc/workflows/utils.py [hunk 7]
```diff
@@ -157,3 +159,33 @@ def slice_wise_fft(in_file, ftmask=None, spike_thres=3., out_prefix=None):
...
+
+    Keyword arguments:
+    name -- workflow name, defaults to UploadWorkflow
+
+    Returns:
+    workflow with inputnode and UploadMetrics node.
+    """
+    from mriqc.reports import upload_qc_metrics
+
+    no_sub =  settings.get('no_sub',False)
```
Comment (oesteban) [inline mriqc/workflows/utils.py]: missing space before `False`,
extra space before `settings`

File: mriqc/workflows/utils.py [hunk 8]
```diff
@@ -157,3 +159,33 @@ def slice_wise_fft(in_file, ftmask=None, spike_thres=3., out_prefix=None):
...
+    email = settings.get('email','')
```
Comment (oesteban) [inline mriqc/workflows/utils.py]: missing space after `'email',`

File: mriqc/workflows/utils.py [hunk 9]
```diff
@@ -157,3 +159,33 @@ def slice_wise_fft(in_file, ftmask=None, spike_thres=3., out_prefix=None):
...
+
+    workflow = pe.Workflow(name=name)
+    inputnode = pe.Node(niu.IdentityInterface(fields=['in_iqms']),
+        name='inputnode')
+    upld = pe.Node(niu.Function(
+    input_names=['in_iqms', 'no_sub','email'], output_names=['response'],
```
Comment (oesteban) [inline mriqc/workflows/utils.py]: Please fix indentation,
missing space after `'no_sub',`

File: mriqc/workflows/utils.py [hunk 10]
```diff
@@ -157,3 +159,33 @@ def slice_wise_fft(in_file, ftmask=None, spike_thres=3., out_prefix=None):
...
+    function=upload_qc_metrics), name='UploadMetrics')
+    upld.inputs.email = email
+    upld.inputs.no_sub = no_sub
+
+    workflow.connect([(inputnode,upld,[('in_iqms','in_iqms')])])
```
Comment (oesteban) [inline mriqc/workflows/utils.py]: missing spaces

File: mriqc/workflows/functional.py [hunk 11]
```diff
@@ -165,6 +167,7 @@ def fmri_qc_workflow(dataset, settings, name='funcMRIQC'):
         (iqmswf, repwf, [('outputnode.out_file', 'inputnode.in_iqms'),
                          ('outputnode.out_dvars', 'inputnode.in_dvars'),
                          ('outputnode.outliers', 'inputnode.outliers')]),
+        (iqmswf, upldwf,[('outputnode.out_file','inputnode.in_iqms')]),
```
Comment (oesteban) [inline mriqc/workflows/functional.py]: missing spaces

File: mriqc/workflows/anatomical.py [hunk 12]
```diff
@@ -153,6 +155,7 @@ def anat_qc_workflow(dataset, settings, mod='T1w', name='anatMRIQC'):
         (segment, repwf, [('tissue_class_map', 'inputnode.segmentation')]),
         (iqmswf, repwf, [('outputnode.out_noisefit', 'inputnode.noisefit')]),
         (iqmswf, repwf, [('outputnode.out_file', 'inputnode.in_iqms')]),
+        (iqmswf, upldwf,[('outputnode.out_file','inputnode.in_iqms')]),
```
Comment (oesteban) [inline mriqc/workflows/anatomical.py]: missing spaces

File: mriqc/reports/utils.py [hunk 13]
```diff
@@ -109,6 +109,78 @@ def read_report_snippet(in_file):
...
+        except KeyError:
+            pass
+
+        if email != '':
+            data['email'] = email
+        secret_key = 'ZUsBaabr6PEbav5DKAHIODEnwpwC58oQTJF7KWvDBPUmBIVFFtwOd7lQBdz9r9ulJTR1BtxBDqDuY0owxK6LbLB1u1b64ZkIMd46'
+        headers = {'token':secret_key,"Content-Type":"application/json"}
+        try:
+            r = requests.put("http://34.201.213.252:5000/measurements/upload",
+                             headers=headers,data = dumps(data))
```
Comment (oesteban) [inline mriqc/reports/utils.py]: PEP8 warnings

**Comment by oesteban** [comment]
@Shotgunosine those are integration tests to make sure that measures don't change inadvertently. I'm updating them. Thanks a lot for this contribution. Well done! 

BTW, I would suggest going to https://circleci.com/gh/nih-fmrif/mriqc/edit and switch to ON the option "Auto-cancel redundant builds" under advanced settings. This way tests will go faster. Please feel free to cancel some of the builds I've submitted (we would be interested in keeping just the last one).

**Comment by Shotgunosine** [comment]
@oesteban Thanks for pointing out the style errors and pointing me towards the PEP8 linter, I'm embarrassed to say I didn't know about that already. I think I've resolved all of the style errors now and hopefully this will build correctly.

**Comment by oesteban** [comment]
@Shotgunosine thanks a lot -- Don't worry about the conflicts, I'll fix them.

**Pull request merged**